include/pybind11/cast.h

```diff
@@ -78,20 +81,152 @@ PYBIND11_NOINLINE inline internals &get_internals() {
     }
     return *internals_ptr;
 }
 
+/** Iterator that starts from a python base class and iterators through all pybind-registered bases
+ * of that class.  If not using Python-side multiple inheritance, this iterator iterates at most
+ * once.
+ */
+class type_info_iterator {
+public:
+    // Start iterating registered bases from the given python type
+    explicit type_info_iterator(PyTypeObject *type) {
+        auto const &type_dict = get_internals().registered_types_py;
+        while (type) {
+            auto it = type_dict.find(type);
+            if (it != type_dict.end()) {
+                // Found a type via single inheritance, so this is a one-shot iterator
+                typeinfo = (detail::type_info *) it->second;
+                size = 1;
+                return;
+            }
+
+#if !defined(PYPY_VERSION)
+            auto bases = reinterpret_borrow<tuple>(type->tp_bases);
+#else
+            // Workaround for PyPy issue #2480
+            auto bases = reinterpret_steal<tuple>(PyObject_GetAttrString((PyObject *) type, "__bases__"));
+#endif
+            if (bases && bases.size() > 0) {
+                if (bases.size() > 1) {
+                    // Hit multiple-inheritance: load the full tree
+                    load_mi_type_info(bases);
+                    return;
+                }
+
+                type = (PyTypeObject *) bases[0].ptr();
+            }
+            else
+                type = nullptr; // No bases, and didn't find a registered instance
+        }
+    }
+    // Default constructor gives a past-the-end iterator
+    type_info_iterator() {}
+
+    using value_type = detail::type_info *;
+    using reference = const value_type &;
+    // The current type typeinfo
+    value_type typeinfo = nullptr;
+    // For multiple inheritance python types, this is the full set of base classes found.  This iterator
+    // object just iterates through this vector:
+    std::shared_ptr<std::vector<type_info *>> all;
+    // The number of registered base classes.  This is known immediately: either there are none, just one, or
+    // there are multiple because of multiple inheritance (which are calculated during construction).
+    size_t size = 0;
+    bool operator==(const type_info_iterator &it) { return typeinfo == it.typeinfo; }
+    bool operator!=(const type_info_iterator &it) { return !(*this == it); }
+    reference operator*() { return typeinfo; }
+    reference operator->() { return typeinfo; }
+    type_info_iterator &operator ++() {
+        if (++position >= size)
+            typeinfo = nullptr;
+        else
+            typeinfo = (*all)[position];
+
+        return *this;
+    }
+
+private:
+    // Position when iterating through the vector (MI-only)
+    size_t position = 0;
+    // Loads bases from a multiple inheritance type; takes the first tuple of multiple bases
+    // encountered.
+    void load_mi_type_info(const tuple &init_bases) {
```

Comment:
Is the order of bases important? If not, a recursive solution might be a bit more compact and easier to follow:
```c++
void add_cpp_bases(std::vector<type_info *> &all, handle bases) {
    for (auto base : reinterpret_borrow<tuple>(bases)) {
        if (!PyType_Check(base.ptr())
            continue;

        auto type = (PyTypeObject *) base.ptr();
        auto it = type_dict.find(type);
        if (it != type_dict.end())
            all.push_back((detail::type_info *) it->second);
        else if (type->tp_bases)
            add_cpp_bases(all, type->tp_bases);
    }
}
```

```c++
std::vector<type_info *> all;
add_cpp_bases(all, type->tp_bases);

std::sort(all.begin(), all.end());
all.erase(std::unique(all.begin(), all.end()), all.end());
```